zcash_client_sqlite/src/testing/pool.rs

```diff
@@ -301,167 +302,456 @@ pub(crate) fn send_single_step_proposed_transfer<T: ShieldedPoolTester>() {
+            if acct == account_id && bad_index == expected_bad_index);
+    };
+
+    let next_reserved = reservation_should_succeed(&mut st, 1);
+    assert_eq!(next_reserved[0], known_addrs[11]);
+
+    // Calling `reserve_next_n_ephemeral_addresses(account_id, 1)` will have advanced
+    // the start of the gap to index 12. This also tests the `index_range` parameter.
+    let newer_known_addrs = st
+        .wallet()
+        .get_known_ephemeral_addresses(account_id, Some(5..100))
+        .unwrap();
+    assert_eq!(newer_known_addrs.len(), (GAP_LIMIT as usize) + 12 - 5);
+    assert!(newer_known_addrs.starts_with(&new_known_addrs[5..]));
+
+    // None of the five transactions created above (two from each proposal and the
+    // one built manually) have been mined yet. So, the range of address indices
+    // that are safe to reserve is still 0..20, and we have already reserved 12
+    // addresses, so trying to reserve another 9 should fail.
+    reservation_should_fail(&mut st, 9, 20);
+    reservation_should_succeed(&mut st, 8);
+    reservation_should_fail(&mut st, 1, 20);
+
+    // Now mine the transaction with the ephemeral output at index 1.
+    // We already reserved 20 addresses, so this should allow 2 more (..22).
+    // It does not matter that the transaction with ephemeral output at index 0
+    // remains unmined.
+    let (h, _) = st.generate_next_block_including(txids1.head);
+    st.scan_cached_blocks(h, 1);
+    reservation_should_succeed(&mut st, 2);
+    reservation_should_fail(&mut st, 1, 22);
+
+    // Mining the transaction with the ephemeral output at index 0 at this point
+    // should make no difference.
+    let (h, _) = st.generate_next_block_including(txids0.head);
+    st.scan_cached_blocks(h, 1);
+    reservation_should_fail(&mut st, 1, 22);
+
+    // Now mine the transaction with the ephemeral output at index 10.
+    let tx = build_result.transaction();
+    let tx_index = 1;
+    let (h, _) = st.generate_next_block_from_tx(tx_index, tx);
+    st.scan_cached_blocks(h, 1);
+
+    // The rest of this test would currently fail without the explicit call to
+    // `put_tx_meta` below. Ideally the above `scan_cached_blocks` would be
+    // sufficient, but it does not detect the transaction as interesting to the
+    // wallet. If a transaction is in the database with a null `mined_height`,
+    // as in this case, its `mined_height` will remain null unless `put_tx_meta`
+    // is called on it. Normally `put_tx_meta` would be called via `put_blocks`
+    // as a result of scanning, but that won't happen for any fully transparent
+    // transaction, and currently it also will not happen for a partially shielded
+    // transaction unless it is interesting to the wallet for another reason.
+    // Therefore we will not currently detect either collisions with uses of
+    // ephemeral outputs by other wallets, or refunds of funds sent to TEX
+    // addresses. (#1354, #1379)
```

Comment:
I banged my head against a brick wall for most of a day trying to fix this, but it isn't easily fixable without doing the implementation of #1354 that we decided is out-of-scope.